File: scripts/ai-workflow/core/content_parser.py
"""Claude APIベースのコンテンツ解析モジュール

正規表現による脆弱な判定を、Claude Messages APIによる
自然言語理解ベースの高精度な抽出に置き換える。

使用例:
    >>> from core.content_parser import ClaudeContentParser
    >>> parser = ClaudeContentParser()
    >>> decisions = parser.extract_design_decisions(planning_content)
    >>> print(decisions)
    {'implementation_strategy': 'CREATE', 'test_strategy': 'UNIT_INTEGRATION', ...}
"""
import os
import json
import anthropic
from pathlib import Path
from typing import Dict, Any, List, Optional


class ClaudeContentParser:
    """Claude Messages APIを使用したコンテンツ解析"""

    def __init__(self, api_key: Optional[str] = None, model: str = "claude-sonnet-4-20250514"):
        """
        初期化

        Args:
            api_key: Anthropic API Key（省略時は環境変数から取得）
            model: 使用するClaudeモデル（デフォルト: claude-sonnet-4-20250514）

        Raises:
            ValueError: API キーが指定されず、環境変数も設定されていない場合
        """
        # API キーの優先順位:
        # 1. 引数で明示的に指定されたキー
        # 2. ANTHROPIC_API_KEY 環境変数
        #
        # 注意: CLAUDE_CODE_OAUTH_TOKEN (OAuth トークン) は使用できません。
        # ClaudeContentParser は anthropic.Anthropic() クライアントを使用しており、
        # これは Anthropic API Key (sk-ant-api03-...) を必要とします。
        resolved_api_key = api_key or os.environ.get("ANTHROPIC_API_KEY")

        if not resolved_api_key:
            raise ValueError(
                "Anthropic API キーが設定されていません。以下のいずれかの方法で設定してください:\n"
                "1. ClaudeContentParser(api_key='your-key') で明示的に指定\n"
                "2. 環境変数 ANTHROPIC_API_KEY を設定\n"
                "\n"
                "注意: CLAUDE_CODE_OAUTH_TOKEN (OAuth トークン) は使用できません。\n"
                "Anthropic API Key は https://console.anthropic.com/settings/keys で取得できます。"
            )

        self.client = anthropic.Anthropic(api_key=resolved_api_key)
        self.model = model

        # プロンプトディレクトリのパスを設定
        self.prompt_dir = Path(__file__).parent.parent / 'prompts' / 'content_parser'

    def _load_prompt(self, prompt_name: str) -> str:
        """
        プロンプトファイルを読み込み

        Args:
            prompt_name: プロンプトファイル名（拡張子なし）

        Returns:
            str: プロンプトの内容
        """
        prompt_file = self.prompt_dir / f'{prompt_name}.txt'
        return prompt_file.read_text(encoding='utf-8')

    def extract_design_decisions(self, document_content: str) -> Dict[str, str]:
        """
        設計書/計画書から戦略判断を抽出

        Args:
            document_content: 設計書または計画書の内容（Markdown形式）

        Returns:
            Dict[str, str]: 戦略判断
                - implementation_strategy: CREATE/EXTEND/REFACTOR
                - test_strategy: UNIT_ONLY/INTEGRATION_ONLY/BDD_ONLY/UNIT_INTEGRATION/UNIT_BDD/INTEGRATION_BDD/ALL
                - test_code_strategy: EXTEND_TEST/CREATE_TEST/BOTH_TEST

        Example:
            >>> parser = ClaudeContentParser()
            >>> doc = '''
            ... ## 実装戦略: CREATE
            ... 新規機能として実装します。
            ... ## テスト戦略: UNIT_INTEGRATION
            ... ユニットテストと統合テストを実施します。
            ... ## テストコード戦略: CREATE_TEST
            ... 新規テストコードを作成します。
            ... '''
            >>> decisions = parser.extract_design_decisions(doc)
            >>> decisions['implementation_strategy']
            'CREATE'
        """
        # プロンプトファイルを読み込み
        prompt_template = self._load_prompt('extract_design_decisions')
        prompt = prompt_template.replace('{document_content}', document_content)

        try:
            response = self.client.messages.create(
                model=self.model,
                max_tokens=1024,
                messages=[
                    {"role": "user", "content": prompt}
                ]
            )

            # レスポンスからテキストを取得
            response_text = response.content[0].text.strip()

            # JSONをパース
            decisions = json.loads(response_text)

            # null値を除外
            return {k: v for k, v in decisions.items() if v is not None}

        except json.JSONDecodeError as e:
            print(f"[WARNING] 戦略判断の抽出に失敗しました（JSONパースエラー）: {e}")
            print(f"[DEBUG] Response: {response_text}")
            return {}
        except Exception as e:
            print(f"[WARNING] 戦略判断の抽出に失敗しました: {e}")
            return {}

    def parse_review_result(self, messages: List[str]) -> Dict[str, Any]:
        """
        Claude Agent SDKのレスポンスメッセージからレビュー結果を抽出

        Args:
            messages: Claude Agent SDKからのレスポンスメッセージ

        Returns:
            Dict[str, Any]: レビュー結果
                - result: str - PASS/PASS_WITH_SUGGESTIONS/FAIL
                - feedback: str - フィードバック全文
                - suggestions: List[str] - 改善提案一覧

        Example:
            >>> parser = ClaudeContentParser()
            >>> messages = ["AssistantMessage(TextBlock(text='**判定: PASS**\\n\\n素晴らしい実装です。'))"]
            >>> result = parser.parse_review_result(messages)
            >>> result['result']
            'PASS'
        """
        # AssistantMessageのTextBlockとResultMessageのresultを抽出
        text_blocks = []
        import re

        for message in messages:
            # ResultMessageのresultフィールドを優先的に抽出
            if 'ResultMessage' in message and 'result="' in message:
                result_start = message.find('result="') + 8
                # 最後の"を見つける（エスケープされた\"は除外）
                result_end = -1
                i = result_start
                while i < len(message):
                    if message[i] == '"' and (i == result_start or message[i-1] != '\\'):
                        result_end = i
                        break
                    i += 1

                if result_end != -1:
                    result_content = message[result_start:result_end]
                    # エスケープシーケンスを置換
                    result_content = result_content.replace('\\n', '\n')
                    result_content = result_content.replace('\\t', '\t')
                    result_content = result_content.replace('\\r', '\r')
                    result_content = result_content.replace('\\"', '"')
                    result_content = result_content.replace("\\'", "'")
                    result_content = result_content.replace('\\\\', '\\')

                    # ResultMessageのresultは完全なレビュー結果なので、そのまま使用
                    text_blocks.append(result_content)
                    continue

            # AssistantMessageのTextBlockを抽出（従来の処理）
            if 'AssistantMessage' in message and 'TextBlock(text=' in message:
                text_start = message.find('TextBlock(text=') + 16
                text_end = message.find('\')', text_start)
                if text_end == -1:
                    continue

                text_content = message[text_start:text_end]

                # エスケープシーケンスを置換
                text_content = text_content.replace('\\n', '\n')
                text_content = text_content.replace('\\t', '\t')
                text_content = text_content.replace('\\r', '\r')
                text_content = text_content.replace("\\'", "'")
                text_content = text_content.replace('\\\\', '\\')

                # デバッグメッセージや前置きを除外
                skip_patterns = [
                    r"^\s*'\s+in\s+message:",
                    r"^\s*\d+→",
                    r"^I'll\s+conduct",
                    r"^Let me\s+",
                    r"^Now\s+let\s+me",
                    r"^Based on\s+my\s+.*review.*,\s*let me\s+provide",
                ]

                should_skip = False
                for skip_pattern in skip_patterns:
                    if re.match(skip_pattern, text_content.strip(), re.IGNORECASE):
                        should_skip = True
                        break

                # 短すぎるメッセージも除外（ただし判定キーワードが含まれている場合は除外しない）
                if len(text_content.strip()) < 50 and '**判定:' not in text_content:
                    should_skip = True

                if not should_skip:
                    text_blocks.append(text_content)

        # テキストブロックを結合
        full_text = "\n".join(text_blocks)

        if not full_text.strip():
            return {
                'result': 'FAIL',
                'feedback': 'レビュー結果が空でした。',
                'suggestions': ['レビュープロンプトを確認してください。']
            }

        # プロンプトファイルを読み込み
        prompt_template = self._load_prompt('parse_review_result')
        prompt = prompt_template.replace('{full_text}', full_text)

        try:
            response = self.client.messages.create(
                model=self.model,
                max_tokens=256,
                messages=[
                    {"role": "user", "content": prompt}
                ]
            )

            # レスポンスからテキストを取得
            response_text = response.content[0].text.strip()

            # JSONをパース
            parsed = json.loads(response_text)
            result_value = parsed.get('result', 'FAIL').upper()

            return {
                'result': result_value,
                'feedback': full_text.strip(),
                'suggestions': []
            }

        except json.JSONDecodeError as e:
            print(f"[WARNING] レビュー結果の抽出に失敗しました（JSONパースエラー）: {e}")
            print(f"[DEBUG] Response: {response_text}")
            # フォールバック: テキストに'PASS'が含まれているか確認
            if 'PASS' in full_text.upper():
                if 'PASS_WITH_SUGGESTIONS' in full_text.upper():
                    result_value = 'PASS_WITH_SUGGESTIONS'
                else:
                    result_value = 'PASS'
            else:
                result_value = 'FAIL'

            return {
                'result': result_value,
                'feedback': full_text.strip(),
                'suggestions': []
            }
        except Exception as e:
            print(f"[WARNING] レビュー結果の抽出に失敗しました: {e}")
            return {
                'result': 'FAIL',
                'feedback': f'レビュー結果の抽出中にエラーが発生しました: {str(e)}\n\n{full_text[:500]}',
                'suggestions': ['エラーログを確認してください。']
            }

File: scripts/ai-workflow-v2/src/core/content-parser.ts
import fs from 'fs-extra';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { OpenAI } from 'openai';
import type { EvaluationDecisionResult, PhaseName, RemainingTask } from '../types.js';

interface ReviewParseResult {
  result: string;
  feedback: string;
  suggestions: string[];
}

export class ContentParser {
  private readonly client: OpenAI;
  private readonly model: string;
  private readonly promptDir: string;

  constructor(options: { apiKey?: string; model?: string } = {}) {
    const apiKey = options.apiKey ?? process.env.OPENAI_API_KEY;
    if (!apiKey) {
      throw new Error(
        [
          'OpenAI API key is required.',
          'Set the OPENAI_API_KEY environment variable or pass apiKey via constructor.',
          'You can create an API key from https://platform.openai.com/api-keys',
        ].join('\n'),
      );
    }

    this.client = new OpenAI({ apiKey });
    this.model = options.model ?? 'gpt-4o-mini';

    const moduleDir = path.dirname(fileURLToPath(import.meta.url));
    this.promptDir = path.resolve(moduleDir, '..', 'prompts', 'content_parser');
  }

  private loadPrompt(promptName: string): string {
    const promptPath = path.join(this.promptDir, `${promptName}.txt`);
    if (!fs.existsSync(promptPath)) {
      throw new Error(`Prompt file not found: ${promptPath}`);
    }
    return fs.readFileSync(promptPath, 'utf-8');
  }

  public async extractDesignDecisions(documentContent: string): Promise<Record<string, string>> {
    const template = this.loadPrompt('extract_design_decisions');
    const prompt = template.replace('{document_content}', documentContent);

    try {
      const response = await this.client.chat.completions.create({
        model: this.model,
        messages: [{ role: 'user', content: prompt }],
        max_tokens: 1024,
        temperature: 0,
      });

      const content = response.choices?.[0]?.message?.content ?? '{}';
      const parsed = JSON.parse(content) as Record<string, string | null | undefined>;
      const result: Record<string, string> = {};
      for (const [key, value] of Object.entries(parsed)) {
        if (typeof value === 'string' && value.trim().length > 0) {
          result[key] = value;
        }
      }
      return result;
    } catch (error) {
      const message = (error as Error).message ?? String(error);
      console.warn(`[WARNING] Failed to extract design decisions: ${message}`);
      return {};
    }
  }

  public async parseReviewResult(messages: string[]): Promise<ReviewParseResult> {
    const textBlocks: string[] = [];

    for (const rawMessage of messages) {
      try {
        const message = JSON.parse(rawMessage);

        // Extract text from assistant messages
        if (message.type === 'assistant' && message.message?.content) {
          for (const block of message.message.content) {
            if (block.type === 'text' && block.text) {
              textBlocks.push(block.text);
            }
          }
        }

        // Extract text from result messages
        if (message.type === 'result' && message.result) {
          textBlocks.push(message.result);
        }

        // Extract text from Codex agent_message items
        if (message.type === 'item.completed' && message.item) {
          const item = message.item as Record<string, unknown>;
          const itemType = typeof item.type === 'string' ? item.type : '';
          if (itemType === 'agent_message') {
            const text = typeof item.text === 'string' ? item.text : '';
            if (text.trim()) {
              textBlocks.push(text);
            }
          }
        }
      } catch (parseError) {
        // Not JSON, try legacy Python-style parsing
        const message = rawMessage ?? '';
        const resultRegex = /result="([^"]*)"/;

        const resultMatch = message.includes('ResultMessage')
          ? message.match(resultRegex)
          : null;
        if (resultMatch) {
          const normalized = this.normalizeEscapedText(resultMatch[1]);
          textBlocks.push(normalized);
          continue;
        }

        if (message.includes('AssistantMessage') && message.includes('TextBlock(text=')) {
          const start = message.indexOf('TextBlock(text=') + 'TextBlock(text='.length;
          const end = message.indexOf("')", start);
          if (end === -1) {
            continue;
          }

          const extracted = this.normalizeEscapedText(message.slice(start, end));
          if (this.shouldKeepAssistantText(extracted)) {
            textBlocks.push(extracted);
          }
        }
      }
    }

    const fullText = textBlocks.join('\n').trim();
    if (!fullText) {
      return {
        result: 'FAIL',
        feedback: 'レビュー結果を解析できませんでした。',
        suggestions: ['レビュー用のプロンプトや実行ログを確認してください。'],
      };
    }

    const template = this.loadPrompt('parse_review_result');
    const prompt = template.replace('{full_text}', fullText);

    try {
      const response = await this.client.chat.completions.create({
        model: this.model,
        messages: [{ role: 'user', content: prompt }],
        max_tokens: 256,
        temperature: 0,
      });

      const content = response.choices?.[0]?.message?.content ?? '{}';
      const parsed = JSON.parse(content) as { result?: string };
      const result = (parsed.result ?? 'FAIL').toUpperCase();

      return {
        result,
        feedback: fullText,
        suggestions: [],
      };
    } catch (error) {
      const message = (error as Error).message ?? String(error);
      console.warn(`[WARNING] Failed to parse review result via OpenAI: ${message}`);

      const upper = fullText.toUpperCase();
      let inferred = 'FAIL';
      if (upper.includes('PASS_WITH_SUGGESTIONS')) {
        inferred = 'PASS_WITH_SUGGESTIONS';
      } else if (upper.includes('PASS')) {
        inferred = 'PASS';
      }

      return {
        result: inferred,
        feedback: fullText,
        suggestions: [],
      };
    }
  }

  private normalizeEscapedText(text: string): string {
    return text
      .replace(/\\n/g, '\n')
      .replace(/\\t/g, '\t')
      .replace(/\\r/g, '\r')
      .replace(/\\"/g, '"')
      .replace(/\\'/g, "'")
      .replace(/\\\\/g, '\\');
  }

  private shouldKeepAssistantText(content: string): boolean {
    const trimmed = content.trim();
    if (!trimmed) {
      return false;
    }

    const skipPatterns = [
      /^\s*'\s+in\s+message:/i,
      /^\s*\d+行/,
      /^I'll\s+conduct/i,
      /^Let me\s+/i,
      /^Now\s+let\s+me/i,
      /^Based on\s+my\s+.*review.*,\s*let me\s+provide/i,
    ];

    if (skipPatterns.some((pattern) => pattern.test(trimmed))) {
      return false;
    }

    if (trimmed.length < 50 && !trimmed.includes('**結果:')) {
      return false;
    }

    return true;
  }

  /**
   * Parse evaluation report and extract decision, remaining tasks, failed phase, and abort reason
   * Uses LLM to extract structured information from natural language
   */
  public async parseEvaluationDecision(content: string): Promise<EvaluationDecisionResult> {
    const template = this.loadPrompt('parse_evaluation_decision');
    const prompt = template.replace('{evaluation_content}', content);

    try {
      const response = await this.client.chat.completions.create({
        model: this.model,
        messages: [{ role: 'user', content: prompt }],
        max_tokens: 2048,
        temperature: 0,
      });

      const responseContent = response.choices?.[0]?.message?.content ?? '{}';
      const parsed = JSON.parse(responseContent) as {
        decision?: string;
        failedPhase?: string | null;
        abortReason?: string | null;
        remainingTasks?: RemainingTask[] | null;
      };

      const decision = (parsed.decision ?? 'PASS').toUpperCase();
      console.info(`[INFO] Extracted decision: ${decision}`);

      // Validate decision
      const validDecisions = ['PASS', 'PASS_WITH_ISSUES', 'ABORT'];
      const isValidFailPhase = decision.startsWith('FAIL_PHASE_');

      if (!validDecisions.includes(decision) && !isValidFailPhase) {
        return {
          success: false,
          decision,
          error: `無効な判定タイプ: ${decision}`,
        };
      }

      // Build result
      const result: EvaluationDecisionResult = {
        success: true,
        decision,
      };

      // Map failedPhase if FAIL_PHASE_*
      if (isValidFailPhase && parsed.failedPhase) {
        const mappedPhase = this.mapPhaseKey(parsed.failedPhase);
        if (!mappedPhase) {
          return {
            success: false,
            decision,
            error: `無効なフェーズ名: ${parsed.failedPhase}`,
          };
        }
        result.failedPhase = mappedPhase;
      }

      // Include abort reason if ABORT
      if (decision === 'ABORT' && parsed.abortReason) {
        result.abortReason = parsed.abortReason;
      }

      // Include remaining tasks if PASS_WITH_ISSUES
      if (decision === 'PASS_WITH_ISSUES' && parsed.remainingTasks) {
        result.remainingTasks = parsed.remainingTasks;
      }

      return result;
    } catch (error) {
      const message = (error as Error).message ?? String(error);
      console.error(`[ERROR] Failed to parse evaluation decision via LLM: ${message}`);

      // Fallback: Try simple pattern matching
      return this.parseEvaluationDecisionFallback(content);
    }
  }

  /**
   * Fallback pattern matching for evaluation decision when LLM parsing fails
   */
  private parseEvaluationDecisionFallback(content: string): EvaluationDecisionResult {
    console.warn('[WARNING] Using fallback pattern matching for evaluation decision.');

    try {
      const patterns = [
        /DECISION:\s*([A-Z_]+)/i,
        /\*\*総合評価\*\*.*?\*\*([A-Z_]+)\*\*/i,
        /(?:判定|決定|結果)[:：]\s*\**([A-Z_]+)\**/i,
      ];

      let match: RegExpMatchArray | null = null;
      for (const pattern of patterns) {
        match = content.match(pattern);
        if (match) break;
      }

      if (!match) {
        console.warn('[WARNING] No decision pattern matched in fallback. Defaulting to PASS.');
        return { success: true, decision: 'PASS' };
      }

      const decision = match[1].trim().toUpperCase();
      console.info(`[INFO] Fallback extracted decision: ${decision}`);

      return { success: true, decision };
    } catch (error) {
      const message = (error as Error).message ?? String(error);
      return {
        success: false,
        error: `判定解析中にエラー: ${message}`,
      };
    }
  }

  private mapPhaseKey(phaseKey: string): PhaseName | null {
    const normalized = phaseKey.toLowerCase().replace(/[-_]/g, '');
    const mapping: Record<string, PhaseName> = {
      planning: 'planning',
      '0': 'planning',
      requirements: 'requirements',
      '1': 'requirements',
      design: 'design',
      '2': 'design',
      testscenario: 'test_scenario',
      '3': 'test_scenario',
      implementation: 'implementation',
      '4': 'implementation',
      testimplementation: 'test_implementation',
      '5': 'test_implementation',
      testing: 'testing',
      '6': 'testing',
      documentation: 'documentation',
      '7': 'documentation',
      report: 'report',
      '8': 'report',
    };

    return mapping[normalized] ?? null;
  }
}
